### PR TITLE
fix(nested-interactive/aria-text): allow "tabindex=-1" on elements with no role

### PR DESCRIPTION
**Pull request by dan-tripp**
Closes issue: #2934

### CODE REVIEW

**Comment by dan-tripp** [comment]
"ci/circleci: test_win — Your tests failed on CircleCI" - I can't tell if that test failed because of my changes, or if it was going to fail anyway.  code owners - let me know if I should do anything.

**Comment by bastilimbach** [comment]
@dan-tripp Thanks for fixing this!
I came across a different case where tabindex="-1" should be ignored: `presentation-role-conflict`

**Fails**:
```javascript
<button>
  Test
  <svg class="icon" tabindex="-1" role="presentation">...</svg>
</button>
```

**Works**:
```javascript
<button>
  Test
  <svg class="icon" role="presentation">...</svg>
</button>
```

**Codepen**: 
https://codepen.io/bastilimbach/full/KKqEVGG

Is this something you are willing to include in this PR? I can also create a separate issue.

**Comment by dan-tripp** [comment]
@bastilimbach I'd be happy to, and that change seems sensible to me.  However, I'm new here, so I think that I had better leave it to the core team to decide what the desired behaviour ought to be.  If it's not too much trouble, could you create a new issue and tag me?  Then I'll try to make a separate PR for your new issue, if it gets some approval from the core team.

**Comment by straker** [comment]
@bastilimbach so with `presentation-role-conflict` that is a true error. The algorithm takes into account any value of tabindex, so testing a similar example all browsers and screen readers ignore the `role=presentation` on the element. 

```html
<ul>
  <li tabindex="-1" role="presentation">World</li>
</ul>
```

Chrome - Treats the li element as a `listitem` role
Firefox - Treats the li element as a `listitem` role
Safari / VO - Announces "group" (instead of no role)

The reason is that because of the tabindex the element can be focused programmatically which means that screen readers must announce a role for the element (it can't be presentational).

**Comment by straker** [comment]
Thanks for taking on this pr. I've been trying to figure out if there were any problems with allowing `tabindex=-1` on the descendants of `role=text`. What I've decided is that we should hold off on doing this. 

The reason for that is that this change directly affects the `nested-interactive` rule, allowing `tabindex=-1` on descendants of interactive elements to pass the rule. [It was pointed out](https://github.com/dequelabs/axe-core/pull/3159#issuecomment-921770698) that doing this would pose problems for screen readers and so we decided to [create new messages to point that out](https://github.com/dequelabs/axe-core/issues/3163). So before we do this we first need to resolve that issue.

If this change only affected the `aria-text` rule I would be fine with it, but since it also affects the `nested-interactive` rule we should hold off.

**Comment by dan-tripp** [comment]
@straker  Many thanks for your comments.  I'll look into that new issue (#3163).

**Comment by straker** [comment]
@dan-tripp I believe this pr is replaced by https://github.com/dequelabs/axe-core/pull/3194, since we don't want tabindex=-1 to pass but now warn of the problem.

**Review by straker — CHANGES_REQUESTED** [review]
Alright. Discussed with @WilcoFiers after thoroughly testing things. Testing both `role="text` and interactive roles (.e.g `role="button"`), we've determined that a tabindex *can be allowed* under certain conditions. Basically if the element with the tabindex is not a widget role, then tabindex on the element is fine. Otherwise we should still flag it as we do now.

So something like this:

```js
// no-focusable-content-evaluate.js
import { getRole, getRoleType } from '../../commons/aria';

function getFocusableDescendants(vNode) {
  if (!vNode.children) {
    if (vNode.props.nodeType === 1) {
      throw new Error('Cannot determine children');
    }

    return [];
  }

  const retVal = [];
  vNode.children.forEach(child => {
    const role = getRole(child);

    if(getRoleType(role) === 'widget' && isFocusable(child)) {
      retVal.push(child);
    } else {
      retVal.push(...getFocusableDescendants(child));
    }
  });
  return retVal;
}
```

**Comment by dan-tripp** [comment]
Right on, I'll try to implement that this week.

**Comment by dan-tripp** [comment]
Sorry I haven't made progress on this.  I'm having a strange git problem and no time to deal with it.  I've updated some tests.  I think I'm close.

**Comment by WilcoFiers** [comment]
@dan-tripp No rush.

**Comment by dan-tripp** [comment]
My latest are ready for review.  Let me know what you think.  And thank you for your patience.

File: lib/checks/keyboard/no-focusable-content-evaluate.js
Comment (WilcoFiers) [inline lib/checks/keyboard/no-focusable-content-evaluate.js:45]:
```suggestion
    if (getRoleType(role) === 'widget' && isFocusable(child)) {
```

File: test/checks/keyboard/no-focusable-content.js
Comment (WilcoFiers) [inline test/checks/keyboard/no-focusable-content.js:39]: Can you add an example that uses span, tabindex and some widget role that shows those failing?

File: test/checks/keyboard/no-focusable-content.js
Comment (WilcoFiers) [inline test/checks/keyboard/no-focusable-content.js:81]: Hmm, that shouldn't have worked. Links don't have a disabled attribute. I opened a separate issue for that (#3315). No need to fix that in this PR. But we should change the example. I suggest:

```suggestion
      '<button id="target"><input value="hello" disabled></button>'
```

**Review by WilcoFiers — CHANGES_REQUESTED** [review]
Close, couple more minor things

**Comment by dan-tripp** [comment]
Okay, I think this PR is ready for review again.  Thank you for all your suggested changes, @WilcoFiers - I think that I'm getting the hang of this.

**Comment by WilcoFiers** [comment]
@dan-tripp Thank you so much for this contribution! Great work, really appreciate it.

**Comment by dan-tripp** [comment]
Great!  Glad I can help.

**Pull request merged**
